src/java.base/share/classes/java/time/format/DateTimeFormatterBuilder.java

```diff
@@ -4975,14 +5027,281 @@ public String toString() {
+         * @return the array index
+         */
+        static long mapToIndex(String type) {
+            return switch (type) {
+                case "am"           -> Calendar.AM;
+                case "pm"           -> Calendar.PM;
+                case "midnight"     -> 2;
+                case "noon"         -> 3;
+                case "morning1"     -> 4;
+                case "morning2"     -> 5;
+                case "afternoon1"   -> 6;
+                case "afternoon2"   -> 7;
+                case "evening1"     -> 8;
+                case "evening2"     -> 9;
+                case "night1"       -> 10;
+                case "night2"       -> 11;
+                default -> throw new InternalError("invalid day period type");
+            };
+        }
+
+        /**
+         * Returns the DayPeriod to array index map for a locale.
+         *
+         * @param locale  the locale, not null
+         * @return the DayPeriod to type index map
+         */
+        static Map<DayPeriod, Long> getDayPeriodMap(Locale locale) {
+            return DAYPERIOD_CACHE.computeIfAbsent(locale, l -> {
+                LocaleResources lr = LocaleProviderAdapter.getResourceBundleBased()
+                        .getLocaleResources(CalendarDataUtility.findRegionOverride(l));
+                String dayPeriodRules = lr.getRules()[1];
+                final Map<DayPeriod, Long> pm = new ConcurrentHashMap<>();
+                Arrays.stream(dayPeriodRules.split(";"))
+                    .forEach(rule -> {
+                        Matcher m = RULE.matcher(rule);
+                        if (m.find()) {
+                            String from = m.group("from");
+                            String to = m.group("to");
+                            long index = DayPeriod.mapToIndex(m.group("type"));
+                            if (to == null) {
+                                to = from;
+                            }
+                            pm.putIfAbsent(
+                                new DayPeriod(
+                                    Long.parseLong(from) * 60,
+                                    Long.parseLong(to) * 60,
+                                        index),
+                                index);
+                        }
+                    });
+
+                // add am/pm
+                pm.putIfAbsent(new DayPeriod(0, 720, 0), 0L);
+                pm.putIfAbsent(new DayPeriod(720, 1_440, 1), 1L);
+                return pm;
+            });
+        }
+
+        /**
+         * Returns the DayPeriod singleton for the locale and index.
+         * @param locale desired locale
+         * @param index resource bundle array index
+         * @return a DayPeriod instance
+         */
+        static DayPeriod ofLocale(Locale locale, long index) {
+            return getDayPeriodMap(locale).keySet().stream()
+                .filter(dp -> dp.getIndex() == index)
+                .findAny()
+                .orElseThrow();
```

Comment:
Isn't is likely that the user could pass in an unexpected `Locale`? If so, then this needs an exception message.

Reply:
Although it will never throw an exception due to unknown locale (am/pm is always available), I added a DateTimeException w/ message supplier to orElseThrow().